src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,181 @@ mod type_keyword {}
 
 #[doc(keyword = "unsafe")]
 //
-/// Code or interfaces whose [memory safety] cannot be verified by the type system.
+/// Code or interfaces whose [memory safety] cannot be verified by the type
+/// system.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// The `unsafe` keyword has two uses: to declare the existence of contracts the
+/// compiler can't check (`unsafe fn` and `unsafe trait`), and to declare that a
+/// programmer has checked that these contracts have been upheld (`unsafe {}`
+/// and `unsafe impl`, but also `unsafe fn` -- see below). They are not mutually
+/// exclusive, as can be seen in `unsafe fn`.
+///
+/// Unsafe Rust adds some capabilities to Safe Rust:
+///
+/// - Dereference [raw pointers]
+/// - Call `unsafe` functions (including foreign functions, some compiler
+/// intrinsics and the raw allocator)
+/// - Implement `unsafe` [`trait`]s
+/// - Mutate [`static`]s (including [`extern`]al ones)
+/// - Access fields of [`union`]s
+///
+/// # Why do we need a keyword for this ?
+///
+/// **No matter what, Safe Rust can't cause Undefined Behavior**. This is
+/// referenced to as [soundness]: a well-typed program actually has the desired
+/// properties. In Safe Rust, the language and the compiler do this mostly by
+/// themselves whereas in Unsafe Rust the programmer must help. The [Nomicon]
+/// has a more detailed explanation on the subject.
+///
+/// # The different meanings of `unsafe`
+///
+/// Depending on the place where it is used, the `unsafe` keyword meaning can
+/// change. See the following [discussion on Rust Internals] for more in-depth
+/// explanations.
+///
+/// Summarizing the main points:
+///
+/// - `unsafe fn`: calling this function means abiding by a contract the
+/// compiler cannot enforce.
+/// - `unsafe trait Trait`: implementing this [`trait`] means abiding by a
+/// contract the compiler cannot enforce.
+/// - `unsafe {}`: the contract necessary to call the operations inside the
+/// block has been checked and is guaranteed to be respected.
+/// - `unsafe impl`: the contract necessary to implement the trait has been
+/// checked and is guaranteed to be respected.
+///
+/// As of now (Rust 1.44), `unsafe fn` also acts like an `unsafe {}` block
+/// around the code inside the function. This means it is not just a signal to
+/// the caller, but also promises that the preconditions for the operations
+/// inside the function are upheld.
+///
+/// Mixing these two meanings can be confusing and [proposal]s exist to use
+/// `unsafe {}` blocks inside such functions when making `unsafe` operations.
+///
+/// # Dereferencing raw pointers
+///
+/// ```rust
+/// let a = 3;
+/// let b = &a as *const _;
+/// // SAFETY: a has not been dropped and references are always aligned,
+/// // so b is a valid address.
+/// let _ = unsafe { *b };
+/// ```
+///
+/// # Marking elements as unsafe
```

Comment:
Oh wait, this is not just explaining the 5 kinds of unsafe capabilities we have. Now I am confused about what the overall structure is. It feels a bit like just a sequence of sections without any clear order or connection, and there are no transitions from one section to the next.

Reply:
Okay I think the "Dereferencing raw pointers" section is the odd one out. If you want to keep it, I suggest moving it to the end and also having a similar (sub)section for union fields, `static mut` etc.

The two sections you have here are about the two sides of `unsafe`, right? If yes, it should connect to the duality established earlier by using the same terminology. E.g. title the section "declaring items unsafe contracts" and "using items with unsafe contracts", or so. You can also have a brief introduction sentence before that saying something like "Next, we take a closer look at these four uses of the `unsafe` keyword, starting with the ones that *define* a contract others have to abide by".

Reply:
Top-level comment incoming :)